docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,236 @@
+      vnet: # specify either 'name' or 'cidr'
+      # name: my-vnet
+        cidr: 10.250.0.0/16
+      workers: 10.250.0.0/19
+status:
+  state: |
+      {
+          "version": 3,
+          "terraform_version": "0.11.14",
+          "serial": 2,
+          "lineage": "3a1e2faa-e7b6-f5f0-5043-368dd8ea6c10",
+          "modules": [
+              {
+              }
+          ]
+          ...
+      }
+```
+
+Extensions which do not require state migration should set `status.state=nil` in their Custom Resources and trigger a normal reconciliation operation if the CR contains the `core.gardener.cloud/operation=restore` annotation.
+
+**Note:** The same size limitations from the previous section are relevant here as well.
+
+### Shoot reconciliation flow changes
+
+The only data which must be stored in the `ShootState` by the Garden Controller Manager is secrets (e.g ca for the apiserver). Therefore the `botanist.DeploySecrets` step is changed. It is split into two functions which take a list of secrets that have to be generated.
+- `botanist.GenerateSecretState` Generates certificate authorities and other secrets which have to be persisted in the ShootState and must not be regenerated on the New Seed.
+- `botanist.DeploySecrets` Takes secret data from the `ShootState`, generates new ones (e.g. client tls certificates from the saved certificate authorities) and deploys everything in the Shoot's control plane on the New Seed
+
+
+### ShootState synchronization controller
+
+The ShootState synchronization controller is part of the garden controller manager. It syncs the state of extension custom resources from the shoot namespace to the garden cluster and updates the corresponding `spec.extension.state` field in the `ShootState` resource. The controller can `watch` Custom Resources used by the extensions and update the `ShootState` only when changes occur.
+
+
+### Migration workflow
+1. Starting migration
+    - Migration can only be started after a Shoot cluster has been successfully created so that the `status.seed` field in the `Shoot` resource has been set
+    - The `Shoot` resource's field `spec.seedName="new-seed"` is edited to hold the name of the __Destination Seed__ and reconciliation is automatically triggered
+    - The Garden Controller Manager checks if the equality between `spec.seedName` and `status.seed`, detects that they are different and triggers migration.
+1. The Garden Controller Manager waits for the __Destination Seed__ to be ready
+    - The Garden Controller Manager lists all deployed extension resources for the Shoot and copies the values in their `status.state` field to the `ShootState` resource
+1. The Shoot State synchronization controller in GCM syncs the `ShootState` from the __Source Seed__ cluster. Later this will be handled by gardenlet (This might not be necessary when using `watches`).
+    - If the __Source Seed's__ API server is not available, then the last successfully synced `ShotoState` is used
```

Comment:
Remove this, we will always use the `ShootState` object. We will have to assume that it's always up to date.